File: superstitioMain/java/superstitio/relics/SuperstitioRelic.java
package superstitio.relics;

import basemod.abstracts.CustomRelic;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.helpers.ImageMaster;
import superstitio.DataManager;
import superstitio.characters.BaseCharacter;
import superstitioapi.utils.updateDescriptionAdvanced;

public abstract class SuperstitioRelic extends CustomRelic implements updateDescriptionAdvanced {
    public static final String DEFAULT_RELIC = "default_relic";
    private Object[] descriptionArgs;

    public SuperstitioRelic(String id, RelicTier relicTier, LandingSound landingSound) {
        super(id, ImageMaster.loadImage(makeImgPath(id)), ImageMaster.loadImage(makeImgPathOutLine(id)), relicTier, landingSound);
        this.largeImg = ImageMaster.loadImage(makeImgPathLarge(id));
    }

    private static String makeImgPath(final String id) {
        return DataManager.makeImgPath(DEFAULT_RELIC, DataManager::makeImgFilesPath_Relic, id);
    }

    private static String makeImgPathOutLine(final String id) {
        return DataManager.makeImgPath(DEFAULT_RELIC, DataManager::makeImgFilesPath_RelicOutline, id);
    }

    private static String makeImgPathLarge(final String id) {
        return DataManager.makeImgPath(DEFAULT_RELIC, DataManager::makeImgFilesPath_RelicLarge, id);
    }

    @Override
    public boolean canSpawn() {
        return AbstractDungeon.player instanceof BaseCharacter;
    }

    @Override
    public final String getUpdatedDescription() {
        this.updateDescriptionArgs();
        String string = getDescriptionStrings();
        string = String.format(string, descriptionArgs);
        return string;
    }

    public int getCounter() {
        return counter;
    }

    @Override
    public final void setDescriptionArgs(Object... args) {
        if (args[0] instanceof Object[])
            descriptionArgs = (Object[]) args[0];
        else
            descriptionArgs = args;
    }

    @Override
    public abstract void updateDescriptionArgs();

    @Override
    public String getDescriptionStrings() {
        return this.DESCRIPTIONS[0];
    }

}

File: superstitioMain/java/superstitio/relics/SuperstitioRelic.kt
package superstitio.relics

import basemod.abstracts.CustomRelic
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.helpers.ImageMaster
import superstitio.DataManager
import superstitio.characters.BaseCharacter
import superstitioapi.utils.UpdateDescriptionAdvanced
import superstitioapi.utils.getFormattedDescription
import java.util.function.BiFunction

abstract class SuperstitioRelic(id: String, relicTier: RelicTier?, landingSound: LandingSound?) : CustomRelic(
    id, ImageMaster.loadImage(
        makeImgPath(id)
    ), ImageMaster.loadImage(makeImgPathOutLine(id)), relicTier, landingSound
), UpdateDescriptionAdvanced {

    init {
        this.largeImg = ImageMaster.loadImage(makeImgPathLarge(id))
    }

    override fun setCounter(counter: Int) {
        super.counter = counter
    }

    fun getCounter():Int{
        return super.counter
    }

    fun updateDescription() {
        this.description = getFormattedDescription()
    }

    override fun canSpawn(): Boolean {
        return AbstractDungeon.player is BaseCharacter
    }

    override fun getUpdatedDescription(): String {
        return this.getFormattedDescription()
    }

    abstract override fun updateDescriptionArgs()

    override fun getDescriptionStrings(): String {
        return DESCRIPTIONS[0]
    }

    override var descriptionArgs: Array<out Any>? = null

    companion object {
        const val DEFAULT_RELIC: String = "default_relic"
        private fun makeImgPath(id: String): String {
            return DataManager.makeImgPath(
                DEFAULT_RELIC,
                BiFunction<String, Array<String>, String>(DataManager::makeImgFilesPath_Relic),
                id
            )
        }

        private fun makeImgPathOutLine(id: String): String {
            return DataManager.makeImgPath(
                DEFAULT_RELIC,
                BiFunction<String, Array<String>, String>(DataManager::makeImgFilesPath_RelicOutline),
                id
            )
        }

        private fun makeImgPathLarge(id: String): String {
            return DataManager.makeImgPath(
                DEFAULT_RELIC,
                BiFunction<String, Array<String>, String>(DataManager::makeImgFilesPath_RelicLarge),
                id
            )
        }
    }
}